File: ground-station/parsing-and-shared-memory/shared_memory.py
#The SharedMemory class is a linked list with that keeps track of data transferred from the arduino
#SharedMemory has a fixed length that is specified to prevent memory overflow
#SharedMemory opens a connection to an InfluxDB to write the data for long term use
class SharedMemory:

	#The node class represents a node in the linked the linked list
	class Node:

		#Constructor: initializes a node in the linked list
		#next (Node): the next node in the linked list
		#previous (Node): the previous node in the linked list
		#data (array: any): the data stored by the node
		def __init__(self, next, previous, data):
			self.next = next
			self.prev = previous
			self.data = data

		#Gets data in a node 
		#Return: (array: any)
		def get_data(self):
			return self.data
		
		#Gets previous node 
		#Return: (Node)
		def get_prev(self):
			return self.prev
		
		#Gets next node
		#Return: (Node)
		def get_next (self):
			return self.next

	#Constructor: initializes a linked list with a specified maximum length
	#max_length (int): the maximum length of the linked list
	#column_names (array: string): the names of the fields in the InfluxDB measurement
	def __init__(self, max_length=None, column_names=None):
		self.max_length = max_length
		self.length = 1
		self.first = self.Node(None, None, column_names) #CONSIDER MAKING COLUMN NAMES NODE ALWAYS LAST
		self.last = self.first

	#Gets first node in linkedList	
	def get_first(self):
		return self.first
	
	#Gets last node in linkedList
	def get_last(self):
		return self.last

	#Adds a new node to the shared memory linked list, if the linked list length exceeds max_length, then remove the last node
	#data (array: any): an array with a row to be added to a CSV sheet
	def write(self, data):
		node_to_add = self.Node(None, self.first, data)
		self.first.next = node_to_add
		self.first = node_to_add
		self.length += 1

		if self.length > self.max_length: #Remove last node if list is too long
			self.last = self.last.next
			self.length -= 1
	
	#Converts one categoryName in the linkedList to an array
	#category_number (int): the index corresponding to the categoryName. 
	#return: (array)
	def convert_to_array(self, category_number):
		returnArray = []
		#We don't want to include the category header node
		currentNode = self.get_last().get_next()

		while currentNode != None:
			returnArray.append(currentNode.get_data()[category_number])
			currentNode = currentNode.get_next()
		
		return returnArray

File: ground-station/groundstation-application-c-sharp/shared_memory.cs
// NOTE: I want to make the next and previous nodes only set by nested classes.
// See: https://stackoverflow.com/questions/42961744/how-to-restrict-access-to-a-nested-class-to-its-container-in-c

using Python.Runtime;

namespace DiagnosticApp;

	/*
* SharedMemory - Specifies an object that multiple threads can access simeltaneously.  Serves as a "shared memory" between them.
* The SharedMemory class is formatted as a linked list, which can be written and read from
*/
public class SharedMemory
{
	/*
	* Node - represents one node in the linked list
	*/
	public class Node<list_data_type>
	{
		public Node<list_data_type>? next {get; set;} //The refernece to the next node on the linked list
		public Node<list_data_type>? prev {get; set;} //The reference to the previous node on the linked list
		public list_data_type[] data {get; private set;} //The list storing the node data
		
		/* 
		* Constructor: intializes a node in the linked list
		* next (Node<T>): the next node in the linked list
		* previous (Node<T>): the previous node in the linked list
		* data (List<T>) : the data stored by the node.  The Generic constraint of the list MUST match that of the node
		* column_names_length (int): the size of the column_names array
		*/
		public Node(Node<list_data_type>? next, Node<list_data_type>? previous, list_data_type[] data)
		{
			this.next = next;
			this.prev = previous;
			this.data = data;
		}
	}

	public int max_length {get; private set;} //the maximum number of nodes on the linked list
	public int length {get; private set;} //the current number of nodes on the linked list
	public Node<double>? first {get; private set;} //the first node on the linked list
	public Node<double>? last {get; private set;} //the last node on the linked list
	public String[] column_names {get; private set;} //the names of the fields in the InfluxDB measurment

	/*
	* Constructor: Initializes the SharedMemory class, which contains a linked list of specified maximum length
	* max_length (int): the maximum length of the linked list.  Precondition: max_length > 0
	* column_names (List<String>): the names of the fields in the InfluxDB measurement
	*/
	public SharedMemory(int max_length, String[] column_names)
	{
		this.max_length = max_length;

		// If max length is less than 1, set it to 1
		if (max_length < 1)
		{
			this.max_length = 1;
		}
		
		this.length = 0;
		this.column_names = column_names;
		this.first = null;
		this.last = null;
	}

	/*
	* Adds a new node to the shared memory linked list.  If the linked list exceeds max_length, then remove the last node
	* data (node_data_type): an array with a row to be added to the database
	*/
	public void write(double[] data)
	{
		if(data.Length != column_names.Length)
		{
			Console.WriteLine("There are more data entries than column names.  Data not added!");
			return;
		}
		Node<double> node_to_add = new Node<double>(null, first, data);

		// Ensure that the first node is NOT null before changing its reference.
		// If the first node is null, then make this node the first and last node.  
		if(first != null)
		{
			node_to_add.prev = first;
			first.next = node_to_add;
			
		}
		else
		{
			last = node_to_add;
		}

		first = node_to_add;
		length += 1;

		// If the list is at its max length, then we want to drop the last node.
		// Currently, we check to make sure the last node is not null, but if max_length > 0, then that should not happen
		if(length > max_length)
		{
			last = last.next;
			last.prev = null;
			length -= 1;
		}

	}

	/*
	* Converts one columtn in the linkedList to an array
	* category_number (int): the index corresponding to the category name (first column is zero)
	*/
	public List<double>? convert_to_array(int category_number)
	{
		if(category_number > column_names.Length)
		{
			return null;
		}

		List<double> return_list = new List<double>();

		Node<double>? current_node = last;

		while(current_node != null)
		{
			return_list.Add(current_node.data[category_number]);
			current_node = current_node.next;
		}

		return return_list;
	}

	/*
	* Overload: Converts one columtn in the linkedList to an array
	* category_name (String): the name of the target category
	*/
	public List<double>? convert_to_array(String category_name)
	{
		for (int i = 0; i < column_names.Length; i++)
		{
			if (column_names[i].Equals(category_name))
			{
				return convert_to_array(i);
			}
		}

		return null;
	}

	/*
	* Converts one columtn in the linkedList to an Python list
	* category_number (int): the index corresponding to the category name (first column is zero)
	*/
	public PyList? convert_to_array_python(int category_number)
	{
		if(category_number > column_names.Length)
		{
			return null;
		}

		PyList return_list = new PyList();

		Node<double>? current_node = last;

		while (current_node != null)
		{
			return_list.Append(new PyFloat(current_node.data[category_number]));
			current_node = current_node.next;
		}

		return return_list;
	}

	/*
	* Overload: Converts one columtn in the linkedList to a Python list
	* category_name (String): the name of the target category
	* return: (PyList)
	*/
	public PyList? convert_to_array_python(String category_name)
	{
		for (int i = 0; i < column_names.Length; i++)
		{
			if (column_names[i].Equals(category_name))
			{
				return convert_to_array_python(i);
			}
		}

		return null;
	}
}

